source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,188 @@
+        [on_complete, end_stream = end_stream_, sz](bool ready) {
+          on_complete(ready, end_stream, sz);
+        },
+        end_stream_);
+  }
+
+private:
+  Buffer::OwnedImpl buffer_;
+  const bool end_stream_;
+};
+
+class CacheInsertChunkTrailers : public CacheInsertChunk {
+public:
+  explicit CacheInsertChunkTrailers(const Http::ResponseTrailerMap& trailers)
+      : trailers_(Http::ResponseTrailerMapImpl::create()) {
+    Http::ResponseTrailerMapImpl::copyFrom(*trailers_, trailers);
+  }
+
+  void send(InsertContext& context,
+            std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) override {
+    // While zero isn't technically true for the size of trailers, it doesn't
+    // matter at this point because watermarks after the stream is complete
+    // aren't useful.
+    context.insertTrailers(*trailers_, [on_complete](bool ready) { on_complete(ready, true, 0); });
+  }
+
+private:
+  std::unique_ptr<Http::ResponseTrailerMap> trailers_;
+};
+
+// We capture the values from encoder_callbacks, and the callbacks as lambdas, so
+// that we don't have to keep checking whether the filter has been deleted.
+// When the filter *is* deleted, we simply replace all the callback lambdas with no-op.
+CacheInsertQueue::CacheInsertQueue(Http::StreamEncoderFilterCallbacks& encoder_callbacks,
+                                   InsertContextPtr insert_context, AbortInsertCallback abort)
+    : dispatcher_(encoder_callbacks.dispatcher()), insert_context_(std::move(insert_context)),
+      low_watermark_bytes_(encoder_callbacks.encoderBufferLimit() / 2),
+      high_watermark_bytes_(encoder_callbacks.encoderBufferLimit()),
+      under_low_watermark_callback_([&encoder_callbacks]() {
+        encoder_callbacks.onEncoderFilterBelowWriteBufferLowWatermark();
+      }),
+      over_high_watermark_callback_([&encoder_callbacks]() {
+        encoder_callbacks.onEncoderFilterAboveWriteBufferHighWatermark();
+      }),
+      abort_callback_(abort) {}
+
+void CacheInsertQueue::insertHeaders(const Http::ResponseHeaderMap& response_headers,
+                                     const ResponseMetadata& metadata, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  // While zero isn't technically true for the size of headers, headers are
+  // typically excluded from the stream buffer limit.
+  insert_context_->insertHeaders(
+      response_headers, metadata,
+      [this, end_stream](bool ready) { onChunkComplete(ready, end_stream, 0); }, end_stream);
+  chunk_in_flight_ = true;
+}
+
+void CacheInsertQueue::insertBody(const Buffer::Instance& chunk, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  if (chunk_in_flight_) {
+    size_t sz = chunk.length();
+    queue_size_bytes_ += sz;
+    chunks_.push_back(std::make_unique<CacheInsertChunkBody>(chunk, end_stream));
+    if (!sent_watermark_ && queue_size_bytes_ > high_watermark_bytes_) {
+      over_high_watermark_callback_();
```

Comment:
in your system you may null out this callback. So don't we need to check for that here?

And if are checkign for that, we could instead just check my proposed OptRef for has_value.

Reply:
This reply is for all three related comments; the reason I don't think it's easier to do it the other way is that I *don't* null out the callback when the filter is destroyed, I set it to a no-op callback.
Which is the point, because it means we don't have to check anything at the callsites.
I can change it if you want, but I think multiple "does the set of callbacks still exist" checks makes it harder to follow to normal flow.

Reply:
ah, this makes sense; i understand why you don't need the conditional here to avoid crashing. But I do wonder whether the logic following the calling of the empty callback makes sense, e.g. setting `sent_watermark_ = true;` even though you have not actually done that.

Maybe it makes no difference, but it's the behavior around these weird timing corner-cases that makes systems behave in interesting ways :)

Maybe worth at least a comment here if not explicit conditionals in the code?

Reply:
Renamed sent_watermark_ to watermarked_ for clarity since it's not necessarily associated with actually sending, and used OptRef rather than lambdas for the standard callbacks.